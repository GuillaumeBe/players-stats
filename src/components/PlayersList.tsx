import { useQuery } from "@apollo/client";

import PLAYERS from "../graphql/queries";
import UserInterface from "../interfaces/player";

export default function Players() {
  const { loading, error, data } = useQuery(PLAYERS);

  if (loading) return <p>Loading...</p>;
  if (error) return <p>Error :(</p>;

  return data.players.map(
    ({
      id,
      firstname,
      lastname,
      sex,
      picture: { url: playerPictureUrl },
      country: {
        picture: { url: countryPictureUrl },
        code,
      },
      stats: { rank, points, weight, height, age },
    }: UserInterface) => (
      <div key={id}>
        <p>
          {firstname}
          {lastname}
          {sex}
          <img src={playerPictureUrl} alt="Player" />
          <img src={countryPictureUrl} alt={code} />
          {rank}
          {points}
          {weight}
          {height}
          {age}
        </p>
      </div>
    )
  );
}
