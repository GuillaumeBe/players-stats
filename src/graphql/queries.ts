import { gql } from "@apollo/client";

const PLAYERS = gql`
  query players {
    players {
      id
      firstname
      lastname
      sex
      picture {
        url
      }
      country {
        picture {
          url
        }
        code
      }
      stats {
        rank
        points
        weight
        height
        age
      }
    }
  }
`;

export default PLAYERS;
