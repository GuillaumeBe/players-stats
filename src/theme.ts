const theme = {
  color: {
    primary: "#4b4a67",
    secondary: "#ddd1c7",
  },
  borderRadius: "5px",
};

export default theme;
