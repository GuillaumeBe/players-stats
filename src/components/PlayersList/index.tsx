import { useQuery } from "@apollo/client";

import Player from "../Player";
import { PLAYERS } from "../../graphql/queries";
import { PlayersInterface } from "../../interfaces/players";

export default function PlayersList() {
  const { loading, data } = useQuery<PlayersInterface>(PLAYERS);

  if (loading) return <div>Loading...</div>;

  const sortedPlayers =
    data &&
    [...data.players].sort((a, b) => (a.stats.rank > b.stats.rank ? 1 : -1));

  return (
    <>
      {sortedPlayers?.map(
        ({ id, firstname, lastname, sex, picture, country, stats }) => (
          <Player
            key={id}
            id={id}
            firstname={firstname}
            lastname={lastname}
            sex={sex}
            picture={picture}
            country={country}
            stats={stats}
          />
        )
      )}
    </>
  );
}
