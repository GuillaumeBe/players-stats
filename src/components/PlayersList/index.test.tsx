import { render, waitFor } from "@testing-library/react";
import { MockedProvider } from "@apollo/client/testing";
import { ThemeProvider } from "styled-components";

import PlayersList from "./index";
import { PLAYERS } from "../../graphql/queries";
import theme from "../../theme";

const successMocks = [
  {
    request: {
      query: PLAYERS,
    },
    result: {
      data: {
        players: [
          {
            id: "player-1",
            firstname: "Stan",
            lastname: "Wawrinka",
            sex: "MAN",
            picture: {
              url: "https://i.eurosport.com/_iss_/person/pp_clubteam/large/325225.jpg",
            },
            country: {
              code: "SUI",
              picture: {
                url: "https://i.eurosport.com/_iss_/geo/country/flag/large/2213.png",
              },
            },
            stats: {
              age: 33,
              rank: 21,
              points: 1784,
              height: 183,
              weight: 81000,
            },
          },
        ],
      },
    },
  },
];

const errorMocks = [
  {
    request: {
      query: PLAYERS,
    },
    error: new Error("An error occurred"),
  },
];

test("renders players list", async () => {
  const { container } = render(
    <ThemeProvider theme={theme}>
      <MockedProvider mocks={successMocks} addTypename={false}>
        <PlayersList />
      </MockedProvider>
    </ThemeProvider>
  );

  await waitFor(() => new Promise((res) => setTimeout(res, 10)));

  expect(container).toMatchSnapshot();
});

test("renders error", async () => {
  const { container } = render(
    <MockedProvider mocks={errorMocks} addTypename={false}>
      <PlayersList />
    </MockedProvider>
  );

  await waitFor(() => new Promise((res) => setTimeout(res, 10)));

  expect(container).toMatchSnapshot();
});

test("renders loading", () => {
  const { container } = render(
    <MockedProvider mocks={successMocks} addTypename={false}>
      <PlayersList />
    </MockedProvider>
  );

  expect(container).toMatchSnapshot();
});
