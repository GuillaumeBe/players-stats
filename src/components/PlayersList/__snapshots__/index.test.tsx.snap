// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders error 1`] = `
<div>
  <div>
    Error
  </div>
</div>
`;

exports[`renders loading 1`] = `
<div>
  <div>
    Loading...
  </div>
</div>
`;

exports[`renders players list 1`] = `
<div>
  <div
    class="sc-bczRLJ eyxgPe"
  >
    <div>
      <div
        class="sc-dkzDqf denMjn"
      >
        Rank: 
        21
      </div>
      <div
        class="sc-hKMtZM lbDXNu"
      >
        Points: 
        1784
      </div>
      <div
        class="sc-eCYdqJ fmJdUt"
      >
        Stan
         
        Wawrinka
        â™‚
        <div
          class="sc-gsnTZi hzCRrq"
        >
          <img
            alt="SUI"
            src="https://i.eurosport.com/_iss_/geo/country/flag/large/2213.png"
          />
        </div>
      </div>
      <ul
        class="sc-gKXOVf MdrjT"
      >
        <li>
          Age: 
          33
        </li>
        <li>
          Weight: 
          81
           kg 
        </li>
        <li>
          Height: 
          1.83
           m
        </li>
      </ul>
    </div>
    <img
      alt="Player"
      class="sc-jSMfEi hdJsHV"
      src="https://i.eurosport.com/_iss_/person/pp_clubteam/large/325225.jpg"
    />
  </div>
</div>
`;
